### ISSUE TITLE
is implementation of mask correct?

### ISSUE THREAD

**Issue by ffmpbgrnn** [vivanov879/translator#3]
I found that you masked prediction, but make states unmasked. I think it is necessary to mask the states when do backpropagation. What do you think?

**Comment by vivanov879** [comment]
What 'states' are you talking about? lstm_h_dec[t], lstm_c_dec[t]?

**Comment by ffmpbgrnn** [comment]
Yes. `c` and `h` are masked in [this implementation](https://github.com/kelvinxu/arctic-captions/blob/master/capgen.py#L445)

**Comment by vivanov879** [comment]
When I coded that part, my logic was that we should mask only those parts of network that can affect the parameters and gradparameters of the neural network. That's why i only masked the decoder output and decoder gradOutput. We didn't mask the states, so we didn't restrict flow of information at all. So, if we mask the encoder input, states of the encoder, states of the decoder, the training might converge faster, but not masking them is still correct. To make the mini-batch work, I appended <UNK> words to make all sentences of equal length, so not masking encoder leads to several <UNK> being fed into network. So the states get polluted with them, that will slow the training. We don't erase states from minibatch to minibatch, so, If we mask the states, we must ensure that the states are correctly read and written.
To sum up, masking them is ok, but should be done carefully. Not masking them is ok. It's my opinion, what do you think?

**Comment by vivanov879** [comment]
Just like reversing the input of the encoder, it will work even without reversing, but it speeds up the training.

**Comment by ffmpbgrnn** [comment]
> my logic was that we should mask only those parts of network that can affect the parameters and gradparameters of the neural network

But encoder input will affect the training process which would affect the parameters, right?

> not masking encoder leads to several being fed into network. So the states get polluted with them, that will slow the training. 

Yes, the states are polluted. I think it would affect the performance not only slow the training.

> Just like reversing the input of the encoder, it will work even without reversing, but it speeds up the training.

You mean reversing the input of the decoder? Actually reversing the order of the decoder will get better performance as stated in Sutskever's paper

Correct me if I am wrong. :-)

**Comment by vivanov879** [comment]
Encoder input will affect parameters, agree.
About performance: i think the best way is to check whether masking states will help.
About reversing: i see he says it improves performance, but i don’t understand why: we insert same information as when not reversed. So, speeds up training, agree. Improves performance, don’t get it. In his paper, he reverses encoder input = ’source sentence’ in his paper.

**Comment by vivanov879** [comment]
Thanks a lot for checking out the project and your very interesting points.

**Comment by ffmpbgrnn** [comment]
Right. He reversed the `encoder` part. Thank **you** for sharing your code :)
